hal/i2c_api.h

```diff
@@ -67,84 +82,162 @@ enum {
 extern "C" {
 #endif
 
+typedef struct {
+    /**< Minimum frequency supported must be set by target device */
+    uint32_t    minimum_frequency;
+    /**< Maximum frequency supported must be set by target device */
+    uint32_t    maximum_frequency;
+    /**< If true, the device can handle I2C slave mode. */
+    bool      supports_slave_mode;
+    /**< If true, supports 10-bit addressing. */
+    bool      supports_10bit_addressing;
+    /**< If true, the device handle multimaster collisions and arbitration safely*/
+    bool      supports_multi_master;
+    /**< If true, supports configuring clock stretching. */
+    bool supports_clock_stretching;
+} i2c_capabilities_t;
+
 /**
  * \defgroup hal_GeneralI2C I2C Configuration Functions
  * @{
  */
 
+/** Fills structure indicating supported features and frequencies on the current
+ *  platform.
+ *
+ * @param[out] capabilities  Capabilities structure filled with supported
+ *                           configurations.
+ */
+void i2c_get_capabilities(i2c_capabilities_t *capabilities);
+
 /** Initialize the I2C peripheral. It sets the default parameters for I2C
- *  peripheral, and configures its specifieds pins.
+ *  peripheral, and configures its pins.
  *
- *  @param obj  The I2C object
- *  @param sda  The sda pin
- *  @param scl  The scl pin
+ *  @param obj       The I2C object
+ *  @param sda       The sda pin
+ *  @param scl       The scl pin
+ *  @param is_slave  Choose whether the peripheral is initialised as master or
+ *                   slave.
  */
-void i2c_init(i2c_t *obj, PinName sda, PinName scl);
+void i2c_init(i2c_t *obj, PinName sda, PinName scl, bool is_slave);
 
-/** Configure the I2C frequency
+/** Release the I2C object.
  *
  *  @param obj The I2C object
- *  @param hz  Frequency in Hz
  */
-void i2c_frequency(i2c_t *obj, int hz);
+void i2c_free(i2c_t *obj);
 
-/** Send START command
+/** Configure the frequency in Hz the I2C peripheral should operate at.
  *
- *  @param obj The I2C object
+ *  @param obj        The I2C object
+ *  @param frequency  Frequency in Hz
+ *
+ *  @returns The actual frequency that the peripheral will be generating to
+ *           allow a user adjust its strategy in case the target cannot be
+ *           reached.
  */
-int  i2c_start(i2c_t *obj);
+uint32_t i2c_frequency(i2c_t *obj, uint32_t frequency);
 
-/** Send STOP command
+/** Enable or disable clock stretching for the I2C peripheral.
  *
- *  @param obj The I2C object
+ *  The behaviour is undefined unless `obj` points to a valid 'i2c_t' object
+ *  and the target supports configuring clock stretching, indicated by the
+ *  'supports_clock_stretching' attribute returned by the 'i2c_get_capabilities'
+ *  function.
+ *
+ * @param obj     The I2C object
+ * @param enabled If 'true' enable clock stretching on the given I2C peripheral,
+ *                otherwise disable it.
  */
-int  i2c_stop(i2c_t *obj);
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
```

Comment:
I think `const` is redundant here

Reply:
The `const` isn't necessary, but I'd rather have things const by default unless it needs to be changed even if it's passed by value.